src/concepts/tail.md

```diff
@@ -1 +1,93 @@
 # Tail Recursion
+In programs, functions often make calls to either themselves (recursive calls) or other functions. There are two types of function calls: non-tail calls, and **tail calls**. A function call is called a **tail call** if the caller does not modify or examine the result of the function call.
+
+If every recursive call made by a function is a **tail call**, that function is called **tail recursive**. Put another way, a function is **tail recursive** if the last operation performed by the function is the recursive call.
+
+Suppose we have the following function to sum the elements of a list:
+```sml
+fun sum [] = 0
+  | sum (x::xs) = x + (sum xs)
+```
+This function is not tail-recursive, because after the recursive call is evaluated, we add the result to `x`. Similarly, if we were to case on `sum xs` or did any other operation on its result, then this function would not be tail-recursive.
+
+In order to write the `sum` function in a tail-recursive manner, we know that the last computation that we can do is the recursive call to `sum`. But, we also need some way to keep track of the sum of the list elements and add to it, since we no longer can add things after the recursive call. We will make an accumulator variable, `acc`, that will keep track of the sum of list elements we have exposed so far, and pass that down through the recursive calls. In our base case, we know we have seen every element in the list and there are no more elements to add, so our base case looks like this:
+```sml
+fun sum' ([], acc) = acc
+```
+In our recursive case, we want to use our accumulator to account for the top element of our list, then pass down that accumulator to the recursive calls. We can do that as follows:
+```sml
+fun sum' (x::xs, acc) = sum' (xs, x + acc)
+```
+Because SML evaluates the function arguments before evaluating the function call, `x + acc` is performed before `sum` is called, and then this updated accumulator value is passed down to the recursive `sum` call on `xs`. Thus, the last operation performed is the recursive call, making this a tail-recursive function.
+
+Putting these two parts together, we have:
```

Comment:
perhaps it could be nice to note that, because of the accumulator element, we must pass in 0 to sum as expected. additionally, since we have changed the type of sum, we can also write the "original" sum function by using this tail-recursive sum that we've written as a helper routine